// postcss-autoreset/dist/createResetRule.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _parser = require('postcss-js/parser');

var _parser2 = _interopRequireDefault(_parser);

var _input = require('postcss/lib/input');

var _input2 = _interopRequireDefault(_input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var source = {
  input: new _input2.default('', { from: 'postcss-autoreset' }),
  start: { line: 1, column: 1 },
  end: { line: 1, column: 1 }
};

function createResetRule(selectors, reset) {
  var selector = selectors.map(function (s) {
    return s.trim();
  }).join(',\n');
  var resetRule = _postcss2.default.rule({
    selector: selector,
    source: source,
    raws: { between: ' ' }
  });
  var root = (0, _parser2.default)(reset);
  root.each(function (node) {
    return node.source = resetRule.source;
  });
  resetRule.append(root);
  return resetRule;
}

exports.default = createResetRule;
module.exports = exports['default'];
